### DISCUSSION TITLE
external dhcp + tftpd-hpa

### DISCUSSION THREAD

**Discussion by rotragit** [ltsp/ltsp#265] [Q&A]
Hi all, I have a simple probably stupid question. 
I have to use an external dhcp and boot the clients with PXE. I've set up tftpd-hpa. The service is running OK (I have looked in the syslog and I see the requests from the client).
With LTSP5 I have used pxelinux.0, so in the dhcp entry the filename attribute was pointing to "ltsp/pxelinux.0" and the default configuration in pxelinux.cfg had the entries pointing to vmlinux and initrd.

Now what filename attribute I have to set so that the client knows what to do to boot using ipxe? (assuming ipxe is the only way to boot the client).

Thank you,
romano

**Comment by alkisg** [comment]
Hi Romano,

the best way is to do this:

- Keep using your external DHCP server
- But **remove** the boot filename and root-path option from its config
- Remove tftpd-hpa from the LTSP server as we prefer dnsmasq
- And finally run `ltsp dnsmasq` on the LTSP server

This will allow the LTSP server to work as proxyDHCP and only send the boot filename and root-path, not IPs. And everything will work out of the box as the LTSP installation page describes, without you having to do anything special.

If you don't want to do that, mention which dhcpd server you're using, because it's not one boot filename but many (for BIOS, UEFI, PC clients, ARM clients etc), so that we point you to configuration instructions for that. isc-dhcp? Windows?

**Comment by rotragit** [comment]
Hi, it's isc-dhcp. I prefer not to use dnsmasq and continue to use tftpd-hpa if that is possible.
Thank you!
romano

**Comment by alkisg** [comment]
Sure, it's easily possible. The configuration for isc-dhcp-server is documented at:

https://ltsp.org/advanced/isc-dhcp-server/

**Comment by rotragit** [comment]
Thank you. I'm not the owner of the isc-dhcp server, so I cannot set the ipxe options in the main configuration. At the same time I cannot use dnsmasq because the clients (+150 clients) are all around the city with many routers and lans. Of course I'm not the network administrator and the routers are configured to propagate dhcp requests only to the institutional dhcp server :-).

So, investigating iPXE a little bit, I understand the best option is to set filename="undionly.kpxe" in the dhcp server host configuration, with a embed script that chain to the dynamic script ltsp.ipxe.

Have you something that does this already? Else I'll go in this direction :-)

**Comment by alkisg** [comment]
Yes, in this case what you describe is the best course.
It's also the best course in pfsense setups, that use isc-dhcp-server but give limited options in the configuration GUI.
Of course then you won't be able to netboot UEFI clients.

I do not have precompiled undionly.kpxe and snponly.efi with "dhcp; chain /ltsp/ltsp.ipxe; boot" embedded script, no.

**Comment by rotragit** [comment]
Thank you, done.
The script is:

```
#!ipxe

dhcp
chain tftp://${next-server}/ltsp/ltsp.ipxe
```

Now the message is "ltsp/image/vmlinux... No such file", but this is a task for Monday :-)

romano

**Comment by alkisg** [comment]
I just found another nice method to "break the iPXE infinite DHCP loop" in Windows or pfSense servers or MikroTik routers, using an undocumented "scriptlet" option, and I documented it in the iPXE forum:

https://forum.ipxe.org/showthread.php?tid=21808&pid=41982#pid41982

I'll include it in ltsp-dnsmasq.conf in the future as it even speeds up booting for non proxyDHCP setups, because it avoids an additional PXE request.

**Comment by maccadia** [comment]
@alkisg nice finding. Do you think it would work with any DHCP server assuming it is able to add an encapsulated DHCP option ? (I have a proprietary DHCP ; unfortunately not ISC nor pfSense).

> it even speeds up booting for non proxyDHCP setups, because it avoids an additional PXE request.

Do you mean an additional DHCP request as you don't use `dhcp` in your scriptlet ?

I tried to use the script `ifopen && chain ltsp/ltsp.ipxe` from an iPXE shell (compiled undionly.kpxe with an embedded script with `shell` inside), but I got `Could not start download: Operation not supported`. Does it work only with a scriptlet ?

**Comment by alkisg** [comment]
I think it should work with any DHCP server that supports encapsulated DHCP options.

The "supported" boot process goes like this:

- PXE BIOS does a first DHCP request which returns e.g. filename="undionly.kpxe"
- PXE BIOS loads and runs undionly.kpxe
- At that point you can use either the scriptlet or an embedded script or Ctrl+B and the command line, to run ifopen/chain without a second DHCP request; and the "config" command should show that there's an ip, a next-server etc
- Otherwise, iPXE does a second DHCP request to get another e.g. filename="menu.ipxe" and run that

So what you've tried should have worked. Try it again with boot.ipxe.org/undionly.kpxe and by pressing Ctrl+B.

Note that this won't work from ipxe.iso, as there's no "PXE BIOS first DHCP request" there.

**Comment by maccadia** [comment]
Thank you for the details. It works when I type the ifopen && chain command with boot.ipxe.org/undionly.kpxe and the one I compile _without an embedded script_. 

If I add an embedded script, it fails with "Could not start download: Operation not supported"
```shell
~/dev/ltsp/ipxe/src$ cat chain_script.ipxe 
#!ipxe

ifopen && chain ltsp/ltsp.ipxe || shell

~/dev/ltsp/ipxe/src$ make bin-x86_64-efi/snponly.efi bin/undionly.kpxe EMBED=chain_script.ipxe 
```

Anyway, it's great to avoid the double DHCP. We had an issue with the second DHCP sometimes failing (?!).

**Comment by maccadia** [comment]
Using ifopen && chain in an embedded script works with this one : 

```shell
#!ipxe

ifopen && chain tftp://${next-server}/ltsp/ltsp.ipxe || shell
```
